Allow PHP version to be changed from Site Settings

danielbachhuber opened the pull request:
Related https://github.com/Automattic/studio/issues/32

## Proposed Changes

Allows the PHP version to be changed from Site Settings:

https://github.com/Automattic/studio/assets/36432/dac23d45-ba44-43cc-9d59-6cc3cd4ea405

Also restarts the server after the PHP version is changed.

## Testing Instructions

1. Create a new site.
2. New site should launch with a default version of PHP 8.0.
3. Navigate to the Studio settings for the site.
4. Edit the PHP version to a different PHP version.
5. Verify the server restarts and the site is running the new PHP version.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

danielbachhuber committed:
Display the PHP version on Site Settings screen

danielbachhuber committed:
Add a modal for editing the PHP version

danielbachhuber committed:
First pass at saving selected PHP version

danielbachhuber commented:
Not correctly saving/applying the PHP version quite yet. My intent is to have the PHP version saved in `appdata-v1.json` first, and then the server restarted from that data. I don't feel like I fully grok the data flow though.

Also need to handle invalid or outdated PHP version stored in `appdata-v1.json` (e.g. what should happen if "5.6" is present?)

Happy to take suggestions on the tests to add.

derekblank commented:
> Not correctly saving/applying the PHP version quite yet. My intent is to have the PHP version saved in appdata-v1.json first, and then the server restarted from that data.

The PHP version is being saved correctly to `appdata-v1.json`, but it looks like the change wasn't being applied to update the site details in the UI (and when starting the server). https://github.com/Automattic/studio/pull/226/ should fix this issue.

> Also need to handle invalid or outdated PHP version stored in appdata-v1.json (e.g. what should happen if "5.6" is present?)

Potentially, we could just fallback to using `DEFAULT_PHP_VERSION` if the  `phpVersion` value in `appdata-v1.json` is not one of the values listed in `const availablePhpVersions = [ '7.4', '8.0', '8.1', '8.2' ];`. Perhaps this array should be stored in constants, too.

derekblank committed:
Update site details to allow user to set PHP version (#226)

* Add phpVersion to updateSiteDetails

* Update tests for changing the PHP version

danielbachhuber committed:
Abstract available PHP versions to a constant

matt-west commented:
Thanks @danielbachhuber!

We had originally planned to add the PHP version picker to the add/edit site form under a new "Advanced settings" section. (Moving the local path field there too.) Clicking the edit link on the site settings screen would launch the full edit site modal.

Was that plan changed? (I may have missed any discussion there.)

I prefer the dedicated modal when editing the PHP version for an existing site. Assuming we're okay with maintaining the additional UI in the app. I think users should be able to set the PHP version when creating a new site too though.

![image](https://github.com/Automattic/studio/assets/417538/d54c34ea-7d4d-42ec-8116-9003e4f4f34f)

danielbachhuber commented:
@matt-west I decided to add this implementation for now because I saw there was a lot of demand for a PHP version switcher and didn't want to go down the rabbit hole adding an overly complex form.

katinthehatsite commented:
> I decided to add this implementation for now because I saw there was a lot of demand for a PHP version switcher

I have not looked at the PR yet but I think it would be good to have PHP version shipped before / around WC EU so that the app could be showcased with this feature. It has been requested and mentioned in multiple demo calls, P2s and conversations and was generally marked as one of the main blockers for not using Studio compared to Local.

matt-west commented:
@danielbachhuber Sounds good. We can add the `Advanced settings` dropdown to the new site form later.

fluiddot commented on src/site-server.ts:
```diff
@@ -99,6 +101,7 @@ export class SiteServer {
 			...this.details,
 			url: this.server.url,
 			port: this.server.options.port,
+			phpVersion: this.server.options.phpVersion || DEFAULT_PHP_VERSION,
```
Nitpick: The  `phpVersion` can be a supported value or `undefined`.
```suggestion
			phpVersion: this.server.options.phpVersion ?? DEFAULT_PHP_VERSION,
```

fluiddot commented on src/constants.ts:
```diff
@@ -20,3 +20,5 @@ export const CLIENT_ID = '95109';
 export const WP_AUTHORIZE_ENDPOINT = 'https://public-api.wordpress.com/oauth2/authorize';
 export const SCOPES = 'global';
+export const DEFAULT_PHP_VERSION = '8.0';
+export const AVAILABLE_PHP_VERSIONS = [ '7.4', '8.0', '8.1', '8.2' ];
```
I wonder if we could obtain the available versions from `wp-now` ([reference](https://github.com/WordPress/wordpress-playground/blob/9fd8cdb9427dd86058752dc4d38577207eed4dec/packages/php-wasm/universal/src/lib/supported-php-versions.ts#L1-L10)). This way we'll provide the same versions offered in Playground.
```
import { SupportedPHPVersions } from '@php-wasm/universal';

...

export const AVAILABLE_PHP_VERSIONS = SupportedPHPVersions;
```

fluiddot replied:
I created https://github.com/Automattic/studio/pull/231 with the proposed changes.

fluiddot reviewed (commented):
I tested the PR and works as expected 🎊 . The only limitation I found is that demo sites don't use the PHP version set. As far as I've tested, they are always created with version `8.1.29`. I wouldn't consider it a blocker but I presume users will expect that the PHP version matches between the local and demo site.

fluiddot committed:
Use PHP constants from `wp-now` (#231)

* Use `DEFAULT_PHP_VERSION` constant from `wp-now`

* Use available PHP versions of Playground

* Add `web-streams-polyfill` package for unit tests

* Polyfill `ReadableStream` in unit tests

Web streams are used by `php-wasm`, so we need to polyfull them if we import the library in unit tests.

danielbachhuber commented:
@Automattic/yolo I think this is ready for a review.

fluiddot commented on src/components/content-tab-settings.tsx:
```diff
@@ -30,6 +32,7 @@ export function ContentTabSettings( { selectedSite }: ContentTabSettingsProps )
 	// Empty strings account for legacy sites lacking a stored password.
 	const storedPassword = decodePassword( selectedSite.adminPassword ?? '' );
 	const password = storedPassword === '' ? 'password' : storedPassword;
+	const phpVersion = useGetPhpVersion( selectedSite );
```
I think the hook `useGetPhpVersion` is not necessary because the PHP version is already coming from the site details in `selectedSite`.
```
	const phpVersion = selectedSite.phpVersion ?? DEFAULT_PHP_VERSION;
```

Not sure if this approach has been followed based on the other hook `useGetWpVersion`. If so, [the WordPress version needs to be retrieved from the WordPress files as it's not referenced in the site details](https://github.com/Automattic/studio/blob/9da80ea926f6df3de22f3dc2497b5f7de66c03e4/src/ipc-handlers.ts#L468-L485). That's the main reason for needing that hook.

fluiddot replied:
I'll add these changes in a separate PR to avoid blocking this one.

fluiddot replied:
> I'll add these changes in a separate PR to avoid blocking this one.

https://github.com/Automattic/studio/pull/239

fluiddot committed:
Add unit tests to cover changing PHP version functionality (#233)

* Mock `matchMedia`

* Expose label in `SettingsRow` component

* Add test to cover the case of changing PHP version

* Revert "Expose label in `SettingsRow` component"

This reverts commit bc37161e54e66ad6a0bdd959f8ef26fc6a0505f3.

* Update change PHP version test case

fluiddot reviewed (approved):
It works as expected 🎊, nice work @danielbachhuber @derekblank   🏅 !

I added some comments that shouldn't block the PR but that I plan to address in a separate PR.

Additionally, I noticed a minor issue regarding the window size after we added this new element. As you can see in the screenshot, the settings tab's content now overflows and causes the content to scroll. If I'm not wrong, we've avoided in the past that this tab have a scroll, so we could consider increasing the minimum window height to do so.

<img width="932" alt="Screenshot 2024-06-12 at 13 42 26" src="https://github.com/Automattic/studio/assets/14905380/e43ebec3-cc0c-4de4-ab98-837729ebe4eb">

cc @matt-west

matt-west commented:
While we've tried to avoid scrolling on the settings screen in the past, we were always going to reach the point that it was needed. I’m happy with letting that screen scroll now.

fluiddot commented:
I wonder if we should warn the user about site restarting when changing the PHP version on a running site. This is not a destructive action but it might be unexpected by the user. WDYT?

matt-west commented:
@fluiddot I'm not sure that's necessary. I feel there's an expectation that changing the PHP version in that modal would apply the change right away.

fluiddot committed:
Remove `getPhpVersion` hook and IPC handler (#239)

* Remove `getPhpVersion` hook and IPC handler

* Use `??` operator instead of `||` when setting the php version

* Add inline comment in Jest setup

The comment clarifies why we need the polyfill.
Related to: https://github.com/Automattic/studio/pull/231#discussion_r1635263097

* Update `ContentTabSettings` unit tests

fluiddot committed:
Bump default PHP version to `8.1` (#240)

* Bump default PHP version to `8.1`

This change is driven by the fact that version `8.0` already reached its end-of-life by 2024.
https://www.php.net/supported-versions.php

* Address failure in `createSite` unit test

wojtekn commented:
> While we've tried to avoid scrolling on the settings screen in the past, we were always going to reach the point that it was needed. I’m happy with letting that screen scroll now.

@matt-west it's okay, but it seems we need to make the scrollbar visible all the time, similarly to the sidebar. Now, the main content sidebar scrollbar appears only if the user starts scrolling.

fluiddot commented:
> @matt-west it's okay, but it seems we need to make the scrollbar visible all the time, similarly to the sidebar. Now, the main content sidebar scrollbar appears only if the user starts scrolling.

FYI I've applied this fix in https://github.com/Automattic/studio/pull/255.

lajennylove commented:
> It works as expected 🎊, nice work @danielbachhuber @derekblank 🏅 !
> 
> I added some comments that shouldn't block the PR but that I plan to address in a separate PR.
> 
> Additionally, I noticed a minor issue regarding the window size after we added this new element. As you can see in the screenshot, the settings tab's content now overflows and causes the content to scroll. If I'm not wrong, we've avoided in the past that this tab have a scroll, so we could consider increasing the minimum window height to do so.
> 
> <img alt="Screenshot 2024-06-12 at 13 42 26" width="932" src="https://private-user-images.githubusercontent.com/14905380/338952293-e43ebec3-cc0c-4de4-ab98-837729ebe4eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2NzM2MjksIm5iZiI6MTcxODY3MzMyOSwicGF0aCI6Ii8xNDkwNTM4MC8zMzg5NTIyOTMtZTQzZWJlYzMtY2MwYy00ZGU0LWFiOTgtODM3NzI5ZWJlNGViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE4VDAxMTUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3ODM5N2EzNGJkYWJiNWJkYWEzMTljMjJiN2ZjNTE3ZmIzMTk0ZjFiMDQ0YWFmMDMyNWQ4NmM5ZjI1NDdlOTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.h0dCAaCMPXLTTG4PIiFWBSsDAt7ZrIOZPL2TlyHzvwk">
> cc @matt-west

Hey there guys!

I don't know if this is related to the new implementation regarding the PHP selector, but as soon as I tried to use it I found permission issues. I tried to chown the permissions but it was having the same issue anyway.

![Screenshot 2024-06-17 at 8 26 10 p m](https://github.com/Automattic/studio/assets/5205775/8aa744b5-71dd-470e-b0e9-9556a63ccaff)

To replicate:
- Switch to PHP 8.2
- Clone this Sage/Vite theme [https://github.com/lajennylove/sage-vite](https://github.com/lajennylove/sage-vite) which requires to do the following
```
composer install
yarn 
yarn build
```
- And then activate the theme, or just live preview it to avoid having issues. 

This is the complete log that I got.
```
Warning: require(/var/www/html/wp-content/cache/acorn/framework/cache/packages.php): Failed to open stream: Permission denied in /var/www/html/wp-content/themes/sage-vite/vendor/illuminate/filesystem/Filesystem.php on line 123

Fatal error: Uncaught Error: Failed opening required '/var/www/html/wp-content/cache/acorn/framework/cache/packages.php' (include_path='.:') in /var/www/html/wp-content/themes/sage-vite/vendor/illuminate/filesystem/Filesystem.php:123 Stack trace: #0 /var/www/html/wp-content/themes/sage-vite/vendor/illuminate/filesystem/Filesystem.php(124): Illuminate\Filesystem\Filesystem::Illuminate\Filesystem\{closure}() #1 /var/www/html/wp-content/themes/sage-vite/vendor/roots/acorn/src/Illuminate/Foundation/PackageManifest.php(111): Illuminate\Filesystem\Filesystem->getRequire('/var/www/html/w...') #2 /var/www/html/wp-content/themes/sage-vite/vendor/roots/acorn/src/Illuminate/Foundation/PackageManifest.php(90): Illuminate\Foundation\PackageManifest->getManifest() #3 /var/www/html/wp-content/themes/sage-vite/vendor/roots/acorn/src/Illuminate/Foundation/PackageManifest.php(79): Illuminate\Foundation\PackageManifest->config('aliases') #4 /var/www/html/wp-content/themes/sage-vite/vendor/roots/acorn/src/Roots/Acorn/Bootstrap/RegisterFacades.php(25): Illuminate\Foundation\PackageManifest->aliases() #5 /var/www/html/wp-content/themes/sage-vite/vendor/roots/acorn/src/Illuminate/Foundation/Application.php(263): Roots\Acorn\Bootstrap\RegisterFacades->bootstrap(Object(Roots\Acorn\Application)) #6 /var/www/html/wp-content/themes/sage-vite/vendor/roots/acorn/src/Illuminate/Foundation/Http/Kernel.php(186): Illuminate\Foundation\Application->bootstrapWith(Array) #7 /var/www/html/wp-content/themes/sage-vite/vendor/roots/acorn/src/Roots/Acorn/Bootloader.php(187): Illuminate\Foundation\Http\Kernel->bootstrap(Object(Illuminate\Http\Request)) #8 /var/www/html/wp-content/themes/sage-vite/vendor/roots/acorn/src/Roots/Acorn/Bootloader.php(103): Roots\Acorn\Bootloader->bootHttp() #9 /var/www/html/wp-content/themes/sage-vite/functions.php(43): Roots\Acorn\Bootloader->boot() #10 /var/www/html/wp-settings.php(663): include('/var/www/html/w...') #11 /var/www/html/wp-config.php(96): require_once('/var/www/html/w...') #12 /var/www/html/wp-load.php(50): require_once('/var/www/html/w...') #13 /var/www/html/wp-admin/admin.php(34): require_once('/var/www/html/w...') #14 /var/www/html/wp-admin/customize.php(13): require_once('/var/www/html/w...') #15 {main} thrown in /var/www/html/wp-content/themes/sage-vite/vendor/illuminate/filesystem/Filesystem.php on line 123
```

I hope this is the right place to add this if not I can move it somewhere else.

Best regards.

danielbachhuber commented:
@lajennylove Can you open a new GitHub issue for the issue? Thanks!

fluiddot commented:
Thanks @lajennylove for reporting the issue. As [Daniel mentioned](https://github.com/Automattic/studio/pull/225#issuecomment-2175893659), it would be great if you could [open a GitHub issue](https://github.com/Automattic/studio/issues/new/choose) with this information. Thanks 🙇 !

lajennylove commented:
Hey @danielbachhuber and @fluiddot I already sent [the report about the issue](https://github.com/Automattic/studio/issues/269) I hope it helps.

Thank you!

fluiddot commented:
> Hey @danielbachhuber and @fluiddot I already sent [the report about the issue](https://github.com/Automattic/studio/issues/269) I hope it helps.
> 
> Thank you!

Great, thank you @lajennylove for reporting the issue.

lajennylove commented:
> > Hey @danielbachhuber and @fluiddot I already sent [the report about the issue](https://github.com/Automattic/studio/issues/269) I hope it helps.
> > Thank you!
> 
> Great, thank you @lajennylove for reporting the issue.

Anytime!
